File: backend/app/api/company.py
"""Company workspace API.

Routes:
  GET /api/company/{ticker}/header   → live quote + identity for the shell header
"""
from fastapi import APIRouter, Depends, Request

from backend.app.models.ticker import Ticker, TickerPath
from backend.app.services.company_snapshot import (
    CompanyFinancials,
    CompanyHeader,
    CompanyOverview,
    build_company_financials,
    build_company_header,
    build_company_overview,
)
from backend.app.services.company_transcripts import (
    Transcript,
    TranscriptList,
    TranscriptSummary,
    build_transcript,
    build_transcript_list,
    summarize_transcript,
)

router = APIRouter(prefix="/company", tags=["company"])


@router.get("/{ticker}/header", response_model=CompanyHeader)
async def get_company_header(
    request: Request, ticker: Ticker = Depends(TickerPath)
) -> CompanyHeader:
    return await build_company_header(request.app.state.fmp, ticker)


@router.get("/{ticker}/overview", response_model=CompanyOverview)
async def get_company_overview(
    request: Request, ticker: Ticker = Depends(TickerPath)
) -> CompanyOverview:
    return await build_company_overview(request.app.state.fmp, ticker)


@router.get("/{ticker}/financials", response_model=CompanyFinancials)
async def get_company_financials(
    request: Request, ticker: Ticker = Depends(TickerPath), period: str = "quarter"
) -> CompanyFinancials:
    period = "annual" if period == "annual" else "quarter"
    return await build_company_financials(request.app.state.fmp, ticker, period=period)


@router.get("/{ticker}/transcripts", response_model=TranscriptList)
async def get_company_transcripts(
    request: Request, ticker: Ticker = Depends(TickerPath)
) -> TranscriptList:
    return await build_transcript_list(request.app.state.fmp, ticker)


@router.get("/{ticker}/transcripts/{year}/{quarter}", response_model=Transcript)
async def get_company_transcript(
    request: Request, year: int, quarter: int, ticker: Ticker = Depends(TickerPath)
) -> Transcript:
    return await build_transcript(request.app.state.fmp, ticker, year, quarter)


@router.post("/{ticker}/transcripts/{year}/{quarter}/summary", response_model=TranscriptSummary)
async def post_company_transcript_summary(
    request: Request, year: int, quarter: int, ticker: Ticker = Depends(TickerPath)
) -> TranscriptSummary:
    md = await summarize_transcript(request.app.state.fmp, ticker, year, quarter)
    return TranscriptSummary(summary_md=md)

File: frontend/lib/api/company.ts
import { apiFetch } from "./core";

// ── Company workspace ─────────────────────────────────────────────────────────

export interface CompanyHeader {
  ticker: string;
  name: string | null;
  exchange: string | null;
  logo_url: string | null;
  currency: string | null;
  price: number | null;
  change: number | null;
  change_pct: number | null;
  delay_label: string;
}

export async function getCompanyHeader(ticker: string): Promise<CompanyHeader> {
  return apiFetch<CompanyHeader>(`/api/company/${encodeURIComponent(ticker)}/header`);
}

export interface StatItem {
  label: string;
  value: number | null;
  // "pct_growth" = growth/CAGR rates: tiny-base artifacts render as "n/m".
  unit: "pct" | "pct_growth" | "x" | "money" | "num" | "int";
}

export interface OverviewStatGroup {
  title: string;
  items: StatItem[];
}

export interface PricePoint {
  date: string;
  close: number;
}

export interface CompanyOverview {
  ticker: string;
  sector: string | null;
  industry: string | null;
  stats: OverviewStatGroup[];
  prices: PricePoint[];
}

export async function getCompanyOverview(ticker: string): Promise<CompanyOverview> {
  return apiFetch<CompanyOverview>(`/api/company/${encodeURIComponent(ticker)}/overview`);
}

export interface CompanyFinancials {
  ticker: string;
  period: "annual" | "quarter";
  periods: string[];
  income: Record<string, (number | null)[]>;
  balance: Record<string, (number | null)[]>;
  cashflow: Record<string, (number | null)[]>;
}

export async function getCompanyFinancials(
  ticker: string,
  period: "annual" | "quarter",
): Promise<CompanyFinancials> {
  return apiFetch<CompanyFinancials>(
    `/api/company/${encodeURIComponent(ticker)}/financials?period=${period}`,
  );
}

export interface TranscriptEvent {
  quarter: number;
  fiscal_year: number;
  date: string;
}

export interface TranscriptList {
  ticker: string;
  events: TranscriptEvent[];
}

export interface TranscriptSegment {
  speaker: string;
  text: string;
}

export interface Transcript {
  ticker: string;
  year: number;
  quarter: number;
  date: string | null;
  segments: TranscriptSegment[];
}

export async function getTranscripts(ticker: string): Promise<TranscriptList> {
  return apiFetch<TranscriptList>(`/api/company/${encodeURIComponent(ticker)}/transcripts`);
}

export async function getTranscript(ticker: string, year: number, quarter: number): Promise<Transcript> {
  return apiFetch<Transcript>(
    `/api/company/${encodeURIComponent(ticker)}/transcripts/${year}/${quarter}`,
  );
}

export async function summarizeTranscript(ticker: string, year: number, quarter: number): Promise<{ summary_md: string }> {
  return apiFetch<{ summary_md: string }>(
    `/api/company/${encodeURIComponent(ticker)}/transcripts/${year}/${quarter}/summary`,
    { method: "POST" },
  );
}
